org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/generators/xml2dtd/XML2DTDGenerator.java

```diff
@@ -0,0 +1,115 @@
+/*******************************************************************************
+* Copyright (c) 2020 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.extensions.generators.xml2dtd;
+
+import java.util.Collection;
+import java.util.Map;
+
+import org.eclipse.lemminx.extensions.generators.AbstractXML2GrammarGenerator;
+import org.eclipse.lemminx.extensions.generators.AttributeDeclaration;
+import org.eclipse.lemminx.extensions.generators.Cardinality;
+import org.eclipse.lemminx.extensions.generators.ElementDeclaration;
+import org.eclipse.lemminx.extensions.generators.Grammar;
+import org.eclipse.lemminx.utils.XMLBuilder;
+
+/**
+ * File Generator implementation to generate DTD from a given XML source.
+ * 
+ */
+public class XML2DTDGenerator extends AbstractXML2GrammarGenerator<DTDGeneratorSettings> {
+
+	@Override
+	protected void generate(Grammar grammar, DTDGeneratorSettings settings, XMLBuilder dtd) {
+		for (ElementDeclaration element : grammar.getElements()) {
+			boolean hasCharacterContent = element.hasCharacterContent();
+			// <!ELEMENT
+			dtd.startDTDElementDecl();
+			dtd.addParameter(element.getName());
+			
+			Collection<ElementDeclaration> children = element.getElements();
+			if (children.isEmpty()) {
+				if (hasCharacterContent) {
+					dtd.addContent(" (#PCDATA)");
+				} else {
+					dtd.addContent(" EMPTY");
+				}
+			} else {
+				dtd.addContent(" (");
+				if (hasCharacterContent) {
+					dtd.addContent("#PCDATA|");
+				}
+
+				boolean sequenced = element.getChildrenProperties().isSequenced();
+				if (sequenced) {
+					boolean first = true;
+					for (Map.Entry<String, Cardinality> elementInfo : element.getChildrenProperties().getCardinalities()
+							.entrySet()) {
+						if (!first) {
+							dtd.addContent(", ");
+						}
+						first = false;
+						dtd.addContent(elementInfo.getKey());
+						Cardinality cardinality = elementInfo.getValue();
+						if (cardinality.getMin() == 0 && cardinality.getMax() == 1) {

```

Comment:
I think we could have something like:
```
if (!hasCharacterContent) {
	if (cardinality.getMin() == 0 && cardinality.getMax() == 1) {
		// ? 0-1
		dtd.addContent("?");
	} else if (cardinality.getMin() == 0 && cardinality.getMax() > 1) {
		// * 0-n
		dtd.addContent("*");
	} else if (cardinality.getMax() > 1) {
		// + 1-n
		dtd.addContent("+");
	}
}
```

because if we add `*` on line 86 of this file, we get this problem:
![image](https://user-images.githubusercontent.com/20326645/85153663-746ba580-b224-11ea-8850-8e96471c8bae.png)

Reply:
fixed with test